File: backend/app/schemas/expense_report.py
"""Pydantic schemas for expense report endpoints."""

from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field, model_validator


class RejectRequest(BaseModel):
    """Request body for POST /reports/{id}/reject."""

    admin_notes: str = Field(..., min_length=1, description="Reason for rejection (required, non-empty)")

    @model_validator(mode="after")
    def validate_admin_notes_not_whitespace(self) -> "RejectRequest":
        if not self.admin_notes.strip():
            raise ValueError("admin_notes must not be blank or whitespace-only")
        return self


class ExpenseReportUpdate(BaseModel):
    """Request body for PUT /reports/{id}. All fields optional."""

    title: Optional[str] = Field(default=None, min_length=1, max_length=255)
    description: Optional[str] = Field(default=None)
    reimbursable_from_client: Optional[bool] = Field(default=None)
    client: Optional[str] = Field(default=None)

    @model_validator(mode="after")
    def validate_client(self) -> "ExpenseReportUpdate":
        if self.reimbursable_from_client and not self.client:
            raise ValueError("client is required when reimbursable_from_client is true")
        if self.client is not None:
            from app.constants import CLIENTS

            if self.client not in CLIENTS:
                raise ValueError(f"client must be one of: {CLIENTS}")
        return self


class AdminExpenseReportUpdate(BaseModel):
    """Request body for PUT /reports/{id} when caller is Admin. All fields optional."""

    title: Optional[str] = Field(default=None, min_length=1, max_length=255)
    description: Optional[str] = Field(default=None)
    reimbursable_from_client: Optional[bool] = Field(default=None)
    client: Optional[str] = Field(default=None)
    admin_notes: Optional[str] = Field(default=None, max_length=1000)

    @model_validator(mode="after")
    def validate_client(self) -> "AdminExpenseReportUpdate":
        if self.reimbursable_from_client and not self.client:
            raise ValueError("client is required when reimbursable_from_client is true")
        if self.client is not None:
            from app.constants import CLIENTS

            if self.client not in CLIENTS:
                raise ValueError(f"client must be one of: {CLIENTS}")
        return self


class StatusAuditLogEntry(BaseModel):
    """Response schema for a single audit log entry."""

    id: int
    expense_report_id: int
    status: str
    changed_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ExpenseReportCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = Field(default=None)
    reimbursable_from_client: bool = Field(default=False)
    client: Optional[str] = Field(default=None)

    @model_validator(mode="after")
    def validate_client(self) -> "ExpenseReportCreate":
        if self.reimbursable_from_client and not self.client:
            raise ValueError("client is required when reimbursable_from_client is true")
        if self.client is not None:
            from app.constants import CLIENTS

            if self.client not in CLIENTS:
                raise ValueError(f"client must be one of: {CLIENTS}")
        return self


class ExpenseReportResponse(BaseModel):
    id: int
    title: str
    description: Optional[str]
    total_amount: float
    status: str
    owner_id: int
    owner_username: str
    created_at: datetime
    reimbursable_from_client: bool
    client: Optional[str]
    admin_notes: Optional[str]

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/expenseReport.ts
/**
 * Expense report types mirroring backend Pydantic schemas.
 */

export interface ExpenseReportCreate {
  title: string;
  description?: string;
  reimbursable_from_client: boolean;
  client?: string;
}

export interface ExpenseReportResponse {
  id: number;
  title: string;
  description: string | null;
  total_amount: number;
  status: string;
  owner_id: number;
  owner_username: string;
  created_at: string;
  reimbursable_from_client: boolean;
  client: string | null;
  admin_notes: string | null;
}

/**
 * Request body for PUT /reports/{id}.
 * All fields are optional — only provided fields are updated.
 */
export interface ExpenseReportUpdate {
  title?: string;
  description?: string;
  reimbursable_from_client?: boolean;
  client?: string;
}

/**
 * A single entry in the status audit log, mirroring the backend StatusAuditLogEntry schema.
 * changed_at is an ISO 8601 UTC string (e.g. "2026-04-23T17:00:00Z").
 */
export interface StatusAuditLogEntry {
  id: number;
  expense_report_id: number;
  status: string;
  changed_at: string;
}

/**
 * Request body for POST /reports/{id}/lines.
 * Creates a new expense line with description, amount, and incurred date.
 */
export interface ExpenseLineCreate {
  description: string;
  amount: number;
  incurred_date: string; // ISO 8601 date: "YYYY-MM-DD"
}

/**
 * Request body for PUT /reports/{id}/lines/{line_id}.
 * All fields are optional — only provided fields are updated.
 */
export interface ExpenseLineUpdate {
  description?: string;
  amount?: number;
  incurred_date?: string; // ISO 8601 date: "YYYY-MM-DD"
}

/**
 * Response body for GET /reports/{id}/lines and related line endpoints.
 * Mirrors the backend ExpenseLineResponse schema.
 */
export interface ExpenseLineResponse {
  id: number;
  report_id: number;
  description: string;
  amount: number;
  incurred_date: string; // ISO 8601 date: "YYYY-MM-DD"
}
